kimchi/src/circuits/polynomials/foreign_field_add/circuitgates.rs

```diff
@@ -0,0 +1,195 @@
+///```text
+/// This module implements foreign field addition.
+///
+///
+/// let a_0, a_1, a_2 be 88-bit limbs of the left element
+/// let b_0, b_1, b_2 be 88-bit limbs of the right element
+/// let m_0, m_1, m_2 be 88-bit limbs of the modulus
+///
+/// Then the limbs of the result are
+/// r_0 = a_0 + b_0 - o * m_0 - 2^88 * c_0
+/// r_1 = a_1 + b_1 - o * m_1 - 2^88 * c_1 + c_0
+/// r_2 = a_2 + b_2 - o * m_2 - 2^88 * ___ + c_1
+///
+/// o = 0 or 1 handles overflows in the field
+/// c_i = -1, 0, 1 are auxiliary variables that handle carries between limbs
+///
+/// We need to do an additional range check to make sure that the result is < the modulus, by
+/// adding 2^(3*88) - m. (This can be computed easily from the limbs of m.) Represent this as limbs
+/// g_0, g_1, g_2.
+/// The upper-bound check can be calculated as
+/// u_2 = r_2 + g_2 + k_1
+/// u_1 = r_1 + g_1 - k_1 * 2^88 + k_0
+/// u_0 = r_0 + g_0 - k_0 * 2^88
+///
+/// k_i = 0 or 1 are auxiliary variables that handle carries between limbs
+///
+/// Then, range check r and o. The range check of o can be skipped if there are multiple additions
+/// and r is an intermediate value that is unused elsewhere (since the final r must have had the
+/// right number of moduluses subtracted along the way).
+///
+/// You could lay this out as a double-width gate, e.g.
+/// a_0 a_1 a_2 b_0 b_1 b_2 o c_0 c_1 ___ k_0 k_1
+/// r_0 r_1 r_2 u_0 u_1 u_2
+///
+///    | col | 'ForeignFieldAdd' | 'Zero'      |
+///    | --- | ----------------- | ----------- |
+///    |   0 | 'a0' (copy)       | 'r0' (copy) |
+///    |   1 | 'a1' (copy)       | 'r1' (copy) |
+///    |   2 | 'a2' (copy)       | 'r2' (copy) |
+///    |   3 | 'b0' (copy)       | 'u0' (copy) |
+///    |   4 | 'b1' (copy)       | 'u1' (copy) |
+///    |   5 | 'b2' (copy)       | 'u2' (copy) |
+///    |   6 | 'o'               |             |
+///    |   7 | 'c0'              |             |
+///    |   8 | 'c1'              |             |
+///    |   9 | TODO: 'c2'        |             |
```

Comment:
Ok, let me show this example. I feel that we may need `c2` for certain large values. I am including some tests to refute or confirm it. But in theory, this was the idea that got me into thinking that we needed one:

- We know limbs are 88-bit length because we range-checked them. This means that in principle, there's no real obstacle to having `a+b` with the most significant limb to all-ones. (Disclaimer: ok, for our particular case that `3*\ell = 264` and `f ≈ 2^256` if a and b are in Ff then the first 8 bits should be zeroes. But in isolation and if we do not check that anyhow, the limb could have all ones).

- Then, adding a2 + b2 would result in 2^89-2 which needs one more bit, thus c2 will be nonzero. 

So, if we want to disregard this case because for our particular scenario c2 should always be zero, then I think it is ok. In any case, removing c2 means one less constraint, so if we can avoid this check because it is implied by others, then I agree.

Reply:
sorry what is `3*\ell`?

Reply:
Oh, \ell is 88, the length of each limb, so I am referring to the maximum bit-value of the three limbs together.